### PR TITLE
Build frontend with Vite

### LINKED ISSUE #17793: Consider replacing Webpack with Vite
### Feature Description

It might be a bit early, but I think it's worth investigating switching the build process to pure esbuild. It supports [experimental code splitting](https://esbuild.github.io/api/#splitting) and has [a plugin ecosystem](https://github.com/esbuild/community-plugins) as well. It should build supremely fast.

Vite was suggested previously as a Webpack replacment but I think it's a dead-end because it relies on legacy systems like Rollup and Babel and seems to be generally header into a similar bloated direction like Webpack. Also, in terms of speed, Vite is barely faster than Webpack in my testing.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by silverwind**
Work in progress, does not fully work yet. TODOs:

- Figure out why dynamic imports fail with `import declarations may only appear at top level of a module`
- Replace `_webpack_public_path` with [renderBuiltUrl](https://vitejs.dev/guide/build.html#advanced-base-options)
- Remove webpack and `webpackChunkName`
- Fix `sortablejs` warning
- Figure out a better file structure for the entry point files
- Enable CSS minify

Interestingly, vite is about 2-3 times as slow as webpack on my machine.

Fixes: https://github.com/go-gitea/gitea/issues/17793

### CODE REVIEW

**Comment by silverwind** [comment]
Still a number of open chunk loading issues. If build performance does not improve, I don't think the switch will be worth it. Even the dev server takes over 1 minute to start up, when webpack does the same in 20s. The majority of build time is spent in rollup.

File: vite.config.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            file: fileURLToPath(new URL('public/js/licenses.txt', import.meta.url)),
+            template(dependencies) {
+              const line = '-'.repeat(80);
+              const goJson = readFileSync('assets/go-licenses.json', 'utf8');
+              const goModules = JSON.parse(goJson).map(({name, licenseText}) => {
+                return {name, body: formatLicenseText(licenseText)};
+              });
+              const jsModules = dependencies.map(({name, version, licenseName, licenseText}) => {
+                return {name, version, licenseName, body: formatLicenseText(licenseText)};
+              });
```
Comment (silverwind) [inline vite.config.js]:
```suggestion
              const jsModules = dependencies.map(({name, version, licenseText}) => {
                return {name, version, body: formatLicenseText(licenseText)};
              });
```

**Comment by silverwind** [comment]
Around 30s of build time is allocated to Monaco, maybe https://github.com/vdesjs/vite-plugin-monaco-editor is needed to bring that down. Or maybe vite has some way to cache specific dependencies.

**Comment by delvh** [comment]
I haven't quite understood yet what the benefit about that would be, especially if it appears to be slower than webpack?
Could you please elaborate why it's a better idea?

**Comment by 6543** [comment]
> Interestingly, vite is about 2-3 times as slow as webpack on my machine.

@silverwind are you on windows?

**Comment by silverwind** [comment]
> I haven't quite understood yet what the benefit about that would be, especially if it appears to be slower than webpack? Could you please elaborate why it's a better idea?

- More modern
- Active development
- Buit-in dev server
- Usually faster

> @silverwind are you on windows?

Mac

**Comment by silverwind** [comment]
So I don't think this will work out. Vite is really hard to integrate when it does not manage the HTML, likely the whole dev server needs a custom implementation as I can't figure out on which path it outputs the assets.

While working on this I found two more issues:

- ` window._globalHandlerErrors` implementation is not strict-mode compatible with this fake array being created.
- jquery.are-you-sure is not strict-mode compatible because it has two accidential globals in the code

The build is still massively slow because of big dependencies Monaco, Swagger-UI and Mermaid. I see no way to speed it up. No idea why it's so much faster in Webpack.

**Comment by silverwind** [comment]
I give up here. Until the speed issue is resolved, which imho can only come from https://github.com/evanw/esbuild/issues/16 and vite adapting esbuild to bundle in place of rollup, but I guess it may take a few years for that to happen.

I will follow up with the discovered issues, specifically the strict mode issue needs to be fixed anyways.

**Comment by 6543** [comment]
well another option and also modern would be https://turbo.build/pack but did not got it working for gitea *jet*

**Comment by silverwind** [comment]
Turbo is definitely on my radar, yes. I think the only thing that we are really missing currently with webpack is `type="module"` output, but it's a minor issue really. Build performance with webpack is alreasy very satisfactory imho.

**Comment by silverwind** [comment]
Maybe the slow build time is acceptable when it is never observed during development. The vite developers have promised to look into the slow prod build times.

I will try to figure out the remaining issues with the dev server, will likely get rid of the `/js/` etc subdirectories. I don't think there is actually missing much to get this into operational state.

**Comment by silverwind** [comment]
Waiting on https://github.com/go-gitea/gitea/pull/25940.

**Comment by silverwind** [comment]
Will revisit this in a future version of Vite that will hopefully have increased the prod build speed.

**Pull request closed without merging**